Port the olympic demo, FilterController, etc.

bshepherdson opened the pull request:
This includes all the various core changes, and some bugs I exposed in the process.

bshepherdson committed:
FilterController: Port the foam.u2.search* family, and the Olympics demo

bshepherdson committed:
MDAO: True is a singleton, not a global

bshepherdson committed:
Mlangs: Bring back part of And.partialEval for FilterController

bshepherdson committed:
Properties: Add fromString property for converting strings

bshepherdson committed:
Element: Set parent state to LOADED after children

bshepherdson committed:
QueryParser: Better handling for Enums

bshepherdson committed:
Mlang: Implement basic partialEval for OR

adamvy-google commented on src/core/FObject.js:
```diff
@@ -440,6 +440,8 @@ foam.CLASS({
       if ( other === this ) return 0;
       if ( ! other        ) return 1;
 
+      if ( ! other ) return 1;
```
Uhh this is just a repeat of the previous if

bshepherdson replied:
Ah, they were added in parallel. I've removed mine.

kgrgreer commented on src/foam/demos/olympics/Controller.js:
```diff
@@ -0,0 +1,156 @@
+        xhr.fromUrl('https://raw.githubusercontent.com/foam-framework/foam/' +
+                    'master/js/foam/demos/olympics/MedalData.json');
+        var self = this;
+        return this.PromisedDAO.create({
```
Do we need to bother with this?  Does it hurt if we just copy data into the dao as it arrives?

bshepherdson replied:
It was making the table re-render and the searches re-run thousands of times, so I figure it doesn't hurt to load all the data first before letting the UI update.

adamvy-google replied:
Shouldn't the table render and search view updates be framed?

adamvy-google replied:
Doing Promises.all for thousands of promises can't be fast.

jacksonic replied:
This would be a case for .eof() returning a promise that optimizes the "wait for full commit" path (if that's definitely required here).

Or if we know the DAO in this case will complete the puts in order, just waiting for the last one would be sufficient.

bshepherdson replied:
Done, simpler and faster now.

kgrgreer commented on src/foam/demos/olympics/Medal.js:
```diff
@@ -0,0 +1,76 @@
+      value: 'Gold',
+      shortName: 'c',
+      aliases: [ 'colour', 'medal' ],
+      tableFormatter: function(medal, e) {
```
In FOAM1/U1 tableFormatters and Views were different but now they're the same, so maybe a better name is in order.

bshepherdson replied:
We've got `view` and now `searchView` - `tableView`?

kgrgreer replied:
Yes, tableView sounds good.

kgrgreer commented on src/foam/demos/olympics/Medal.js:
```diff
@@ -0,0 +1,76 @@
+      tableFormatter: function(medal, e) {
+        return e.E('span').cssClass(medal.color.label).add(medal.color.label);
+      },
+      comparePropertyValues: function(o1, o2) {
```
Can't this be simplified now that the medal colour is an Enum? Also, why is it an enum?

bshepherdson replied:
This can indeed be skipped; `Enum`s sort by their ordinals by default.

I guess the medals didn't need to be an enum, but it's handy to get their labels and now sort order from the enum. Also trying to use enums here has been shaking down bugs in the query parser, filter controller and elsewhere, so it's a useful test in any case.

bshepherdson commented:
A general comment about the full-text search: we don't actually support `aliases` right now. There's a TODO for it in the query parser. Should we add that now(ish)?

bshepherdson committed:
TableView: Rename tableFormatter to tableCellView

kgrgreer commented on src/foam/u2/ChoiceViewTrait.js:
```diff
@@ -0,0 +1,164 @@
+foam.CLASS({
+  package: 'foam.u2',
+  name: 'ChoiceViewTrait',
+
```
Add docs. What is this class for?

bshepherdson replied:
Done.

kgrgreer commented on src/foam/u2/search/Autocompleter.js:
```diff
@@ -0,0 +1,69 @@
+ * limitations under the License.
+ */
+foam.CLASS({
+  package: 'foam.u2.search',
```
Shouldn't we have this functionality in just foam.u2 for people who want autocomplete for purposes other than search?

bshepherdson replied:
Good point. I've moved this.

kgrgreer commented on src/foam/u2/search/FilterController.js:
```diff
@@ -0,0 +1,524 @@
+          .start('span')
+              .cssClass(this.myCls('count-text'))
+              .add(this.countString$)
+              .end()
```
I think starts and ends should be on the same level. This style is difficult to read.

kgrgreer commented on src/foam/u2/search/FilterController.js:
```diff
@@ -0,0 +1,524 @@
+          self.searchMgr_.add(self.search);
+        }
+
+        self.setTimeout(function() { self.searchFields; }, 0);
```
Document why this is needed.

bshepherdson replied:
Done. And actually the setTimeout is not needed anymore, too.

kgrgreer commented on src/foam/u2/search/GroupAutocompleteSearchView.js:
```diff
@@ -0,0 +1,159 @@
+      value: ''
+    },
+    {
+      name: 'viewFactory',
```
Would just 'view' be a better name?

adamvy-google replied:
Should this also be of class: 'foam.u2.ViewSpec'

bshepherdson replied:
Changed to `class: 'foam.u2.ViewSpec'` and `name: 'viewSpec'`.

kgrgreer commented on src/foam/u2/search/GroupAutocompleteSearchView.js:
```diff
@@ -0,0 +1,159 @@
+      }
+    },
+    {
+      class: 'Class2',
```
Can we come up with a better name than Class2?

bshepherdson replied:
Maybe, but that's out of scope for this CL.

kgrgreer commented on src/foam/u2/search/GroupBySearchView.js:
```diff
@@ -0,0 +1,158 @@
+      var exists = typeof choice !== 'undefined' && choice !== '';
+      this.predicate = exists ? this.op$cls.create({
+        arg1: this.property,
+        arg2: this.Constant.create({ value: choice })
```
I wonder if we should handle the adapting in the arg# adapt methods rather than in the external compile() method. Would mLangs easier to use directly.

kgrgreer commented on src/foam/u2/tag/Select.js:
```diff
@@ -20,23 +20,26 @@ foam.CLASS({
   extends: 'foam.u2.View',
 
+  implements: [
+    'foam.u2.ChoiceViewTrait'
```
What's the advantage to being a trait rather than just a View?

bshepherdson replied:
I was following the lead of FOAM 1 here. On the other hand, `ChoiceViewTrait` could have been a superclass rather than a trait in FOAM 1 as well.

Should I move it to be a parent class instead?

kgrgreer replied:
I think so.

adamvy-google replied:
As mentioned for the Input tag changes, should we really be adding this functionality to the Select tag?

Things under foam.u2.tag, are supposed to just represent raw tags, and I feel like more advanced stuff such as making a choice view that accepts choices or choiceDao should be part of a higher level component.

kgrgreer commented on src/foam/u2/tag/Select.js:
```diff
@@ -20,23 +20,26 @@ foam.CLASS({
+        if ( ! this.data ) {
+          this.index = 0;
+        }
+      }.bind(this));
```
Don't bind(), since you have self defined above, just use that.

bshepherdson replied:
Done.

kgrgreer commented on src/foam/u2/tag/Select.js:
```diff
@@ -20,23 +20,26 @@ foam.CLASS({
-      this.attrSlot().linkFrom(this.data$);
+      this.attrSlot().linkFrom(this.index$);
+
+      this.onload.sub(function() {
```
Document what this block does/is-for.

bshepherdson replied:
Done.

kgrgreer commented on src/lib/QueryParser.js:
```diff
@@ -192,7 +192,7 @@ foam.CLASS({
       name: 'grammar_',
       factory: function() {
-        var cls = foam.lookup(this.of);
+        var cls = this.of$cls;
```
Why 'cls' rather than 'class'?

bshepherdson replied:
I didn't write `Class2` for this PR; let's take up the questions of its design in a meeting.

adamvy-google commented on src/foam/demos/olympics/Medal.js:
```diff
@@ -0,0 +1,70 @@
+      class: 'Enum',
+      of: 'foam.demos.olympics.MedalColor',
+      name: 'color',
+      value: 'Gold',
```
I don't think default value work for enums like this yet.  You can do foam.demos.olymipcs.MedalColor.GOLD for now.

bshepherdson replied:
I've actually just removed it; I don't think we were using it anyway.

adamvy-google commented on src/foam/mlang/CannedQuery.js:
```diff
@@ -0,0 +1,30 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+foam.CLASS({
```
Is this class actually used?  If not remove it

bshepherdson committed:
Mlangs: Add GroupBy

bshepherdson committed:
Add NullDAO

bshepherdson committed:
DAOs: Add compareTo for DAOs that only checks identity and UIDs

This prevents broken, expensive comparisons of the internals of different DAOs.

adamvy-google commented on src/foam/u2/tag/Input.js:
```diff
@@ -51,6 +53,21 @@ foam.CLASS({
       if ( this.type ) this.attrs({type: this.type});
       this.cssClass(this.myCls());
       this.link();
+      if ( this.autocompleter ) {
```
Should these things really be added to the Input tag model?

I feel like we should have a higher level TextFieldView that composes an input tag and hooks up autocomplete and what not.  Otherwise it cuts down on the re-usability of the raw tag.

bshepherdson committed:
ChoiceView: Add foam.u2.view.ChoiceView, to handle controller logic

Keep the u2.tag.* classes as simple wrappers for the DOM controls. The complex
logic lives in foam.u2.view.ChoiceView.

bshepherdson committed:
Add foam.u2.view.TextField for the tricky logic; keep foam.u2.tag.Input simple

adamvy-google commented on src/foam/u2/tag/Select.js:
```diff
@@ -37,6 +40,7 @@ foam.CLASS({
       var self = this;
 
       this.attrSlot().linkFrom(this.data$);
+      this.attrs({ size: this.size$ });
```
this.setAttribute('size', this.size$); would save an enumeration over the object, buts its probably pretty minor.

adamvy-google commented on src/foam/u2/tag/Input.js:
```diff
@@ -48,7 +49,8 @@ foam.CLASS({
     function initE() {
-      if ( this.type ) this.attrs({type: this.type});
+      if ( this.type ) this.attrs({ type: this.type });
+      if ( this.list ) this.attrs({ list: this.list$ });
```
Does this need to be reflected into a property?  Or could we just use .attrs()/.setAttribute() from the outside?

bshepherdson replied:
Fixed, this is done from the outside now.
